### ISSUE TITLE
Specifying the TLS version

### ISSUE THREAD

**Issue by mherrmann** [google/omaha#194]
(Continuing a discussion from #193).

As confirmed by @sorinj, Omaha currently uses whichever TLS version is the OS default when connecting to the server. Unfortunately, on Windows 7 this is TLS 1.0, which is very old and insecure.

The OS default can be changed on at least Windows 7 by installing [Windows Update 3140245](https://support.microsoft.com/en-us/help/3140245/update-to-enable-tls-1-1-and-tls-1-2-as-default-secure-protocols-in-wi) and setting the `DefaultSecureProtocols` registry value to `0x00000800`.

This has multiple problems: Changing the OS default affects not just Omaha, but all applications on the user's system. Furthermore, it is unnecessarily complicated. The client has the update server URL hard-coded. If it knows that the server supports a more modern TLS version such as 1.2, then it would be perfectly reasonable for the client to also use this version when connecting to the server.

This issue is thus not really a "bug" in Omaha. The current stance seems to be "use the OS default", and Omaha seems to do this. However, for the reasons described above, I would suggest changing this stance.

A good solution would be to make the TLS version configurable. Furthermore, TLS 1.0 is so old by now that it may make sense to disable it in Omaha entirely.

Because I'm pressed for time for a project, I'd also like to ask here if you can recommend a way of quickly implementing this in the current client code. My idea would have been to set `WINHTTP_OPTION_SECURE_PROTOCOLS` in [SimpleRequest::Connect](https://github.com/google/omaha/blob/1b18ab3989b8cf503d5e46a351c44c9d2754caed/omaha/net/simple_request.cc#L301). Would that be a good solution?

Thank you.

**Comment by sorinj** [comment]
SimpleRequest::Connect would work to implement this change. Thank you for bringing it up.

**Comment by sorinj** [comment]
I'd like to discuss this idea a bit further since I am not convinced Omaha code needs to be changed.

It looks like the intention here is to make Omaha use TLS 1.2 on Windows 7, when TLS 1.2 has been made available by WU3140245. 

The support article says `Applications [...] that use the WINHTTP_OPTION_SECURE_PROTOCOLS flag can't use TLS 1.1 or TLS 1.2 protocols. This is because the definition of this flag doesn't include these applications and services. This update adds support for DefaultSecureProtocols registry entry that allows the system administrator to specify which SSL protocols should be used **when the WINHTTP_OPTION_SECURE_PROTOCOLS flag is used**.`. 

I am trying to say that this does not apply to Omaha since Omaha is not using the `WINHTTP_OPTION_SECURE_PROTOCOLS`. As it stands now, I expect Omaha to use the strongest TLS/SSL stack available on the host, which is going to be TLS 1.2 (including on Windows 7, if the patch was applied, regardless of the admin registry settings.)

It looks to me that the intention of the `DefaultSecureProtocols ` registry value is to fix the behavior of programs who inadvertently used  `WINHTTP_FLAG_SECURE_PROTOCOL_ALL` to mean **all** security protocols where in fact they were limited to only a couple of them in the Windows 7 case:

```
#define WINHTTP_FLAG_SECURE_PROTOCOL_ALL    (WINHTTP_FLAG_SECURE_PROTOCOL_SSL2 |\
                                             WINHTTP_FLAG_SECURE_PROTOCOL_SSL3 |\
                                             WINHTTP_FLAG_SECURE_PROTOCOL_TLS1)
```

**Comment by mherrmann** [comment]
I understand your theoretical analysis. However, its conclusion is not what I see in practice.

My server only accepts TLS 1.2 (and thus refuses TLS 1.0). I have a pretty clean, old Windows 7 VM without Update 3140245 installed. Here's what I see:

* Base configuration: fails
* After installing Windows Update 3140245 and restarting: fails
* After setting `DefaultSecureProtocols` in `HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\Internet Settings\WinHttp`: fails
* After restarting: works!

When I say "fails", then in all of the above cases I mean that the Tagged Meta Installer fails to download and I get the Omaha log entry `[SimpleRequest::Send][0x80072f8f][0]`. According to my analysis, this means `GetLastError` returned `0x2f8f` which is [`ERROR_WINHTTP_SECURE_FAILURE`](https://github.com/Tencent/mars/blob/0ad552ae7cbd8cc3d72945be00abe65644668633/mars/comm/windows/win32/winhttp.h#L1195).

In a strict logical sense, what the [support article](https://support.microsoft.com/en-us/help/3140245/update-to-enable-tls-1-1-and-tls-1-2-as-default-secure-protocols-in-wi) says does not preclude this behavior. It only speaks about apps that use `WINHTTP_OPTION_SECURE_PROTOCOLS`. It does not say what happens for apps that don't use that flag, of which Omaha is one.

So to fully drive the point home: On at least Win 7, installing WU 3140245 does not automatically make WinHttp (and thus Omaha) use the strongest TLS version. It still uses the insecure 1.0. From what I read online (I can't remember where, I'm afraid) it may be that this is to preserve backwards compatibility.

**Comment by mherrmann** [comment]
(For the record: `SimpleRequest::Connect` is actually, I think, not a great place to set `WINHTTP_OPTION_SECURE_PROTOCOLS`. The reason is that this setting is scoped on the _session_, while `SimpleRequest::Connect` just receives the "global" session and is scoped on the request. I think the correct place is `WinHttp::Open`, which is where the session is initialized. This change successfully makes Omaha use TLS 1.2 for us.)

**Comment by sorinj** [comment]
Thank you Michael. What value are you going to set for the WINHTTP_OPTION_SECURE_PROTOCOLS option in your case?

**Comment by mherrmann** [comment]
I use `WINHTTP_FLAG_SECURE_PROTOCOL_TLS1_2` to force TLS 1.2. If you want, I could create a pull request that adds a new constant to `constants.h` that specifies the TLS flag with default `WINHTTP_FLAG_SECURE_PROTOCOL_ALL`.

**Comment by mherrmann** [comment]
I'm very sorry @sorinj but a project caught up with me and I won't be able to send a PR soon. If that changes, I will let you know. But I'm afraid not in the foreseeable future. Sorry.

**Comment by sorinj** [comment]
I landed something here: https://github.com/google/omaha/commit/6476d0718f20d7ecb2f93d72a7bd01e4f35add55

**Comment by mherrmann** [comment]
Okay great! Thank you.

**Comment by sorinj** [comment]
This change (6476d07) is breaking some servers. I am reopening the issue for further investigation.

**Comment by sorinj** [comment]
I landed  6db3d6f3f08bc16da7a7d87d6b125e8e95c97807 which we believe is enabling TLS 1.2 on all platforms starting with Windows 7 (assuming that the host can talk the protocol).

**Comment by sorinj** [comment]
Google is seeing excellent results after testing with https://github.com/google/omaha/commit/6db3d6f3f08bc16da7a7d87d6b125e8e95c97807 
TLS1.2 is mostly used and only few transactions go over TLS1.0
